### PR TITLE
Topic recognition (s46387334): Siamese Network for ISIC 2020 Data Set

### PR DESCRIPTION
**Pull request by timryall**
### Overview
This PR implements a Siamese Network model (and associated components) designed to classify skin lesions from the ISIC 2020 Kaggle Challenge data set as either 'normal' or 'melanoma'. When trained - this model was able to achieve an AUR-ROC of `0.867` on a testing set.

`README.md` details the implementation and results - and how these results can be reproduced in a local environment.

### Introduced Files
- `dataset.py`: This module introduces the `TripletDataGenerator` class which is designed to form (and transform) given data into triples that can be used by `TripletLoss`  during training. It also introduces a number of functions e.g.  `get_isic2020_data`, that are designed to extract and load the ISIC 2020 data set so that it is ready for use by the `SiameseNet` model.

- `modules.py`: This module introduces the `SiameseNet` class which is a Siamese Network based model that uses an 'adapted' `resnet50` for its feature extractor and a single layer neural net (perceptron) for the classifier. The `TripletLoss` class is also introduced and it will be the loss function used to train the feature extractor. This file will also contain the hyper parameter config used for training and testing with `SiameseNet`.

- `train.py`: This module contains the training loop used to train `SiameseNet`. The model will be trained using a set of hyper parameters defined in the config of `modules.py`. Each epoch of training a validation set will be predicted to evaluate the current models preformance.

- `predict.py`:  This module is used to show example usage the trained SiameseNet model by using the model to predict
on a testing set from the ISIC 2020 data set. This should be used after training has taken place. Evaluation metrics and figures can be produced by functions such as `produce_evaluation_metrics` and  `produce_evaluation_figures`.

- `README.md`: This file contains details of the implementation and testing of the model. Figures and metrics from the final model are also provided. Details on how to reproduce the results and use the files in the PR can be found in this file. 

- `readme_figures/*`: This folder contains figures that are used in `README.md`.

### Requirements for using this model
- The dataset must be downloaded from  [This Link](https://www.python.org/downloads/release/python-31014/) as detailed in `README.md`
- Python - 3.10.14 must be installed + the following Python packages:
  - matplotlib - 3.7.5
  - numpy - 1.26.4
  - pandas - 2.2.2
  - scikit-learn - 1.2.2
  - seaborn - 0.12.2
  - torch - 2.4.0
  - torchvision - 0.19.0

### Testing
Please see `README.md` for results of testing and how these testing results were produced.

Timothy Ryall
s46387335

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **20**

Solves problem: The solution is well-suited to the problem (**5**).
The confusion matrix, loss, accuracy, ROC, and AUC curves are provided.
Triplet loss is utilized, which is suitable for the problem.
Class imbalance is addressed through oversampling of malignant cases.

Implementation functions as intended (**3**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:
- Training log shows progressive improvement in metrics with best models saved by validation AUC ROC.
-  The model demonstrates reasonably good sensitivity (76.12%) for melanoma detection, high specificifty (82.76%) and almost the required accuracy (79.44). 

- Training loss decreases, but validation loss fluctuates (possible overfitting) , hyperparameter tuning and regularization for better robustness..

- Validation accuracy instability suggests potential model refinement needs.

- The t-SNE visualization suggests that the triplet loss may not have fully succeeded in effectively separating classes by pulling similar examples closer together and pushing different classes further apart. To enhance cluster separation, consider mining semi-hard or hard triplets. it potentially leads to significantly improved separation in the embedding space."

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Great work. Remove gitignore file for me to merge, doesnt affect grade
